draft-ietf-tls-tls13.md

```diff
@@ -2357,84 +2372,109 @@ extension type (Supported Group Extension):
+                   KeyShare client_offers<4..2^16-1>;
 
-offers
-: A list of ClientKeyShareOffer values in descending order of
-  client preference.
-{:br }
+               case server:
+                   KeyShare server_offer;
+           }
+       } KeyShareOffer;
+
+client_offers
+: A list of offered KeyShare values in descending order of client preference.
+  This vector MUST NOT be empty. Clients not providing a KeyShare MUST instead
+  omit this extension from the ClientHello.
 
-Clients may offer an arbitrary number of ClientKeyShareOffer
-values, each representing a single set of key exchange parameters;
-for instance a client might offer shares for several elliptic curves
-or multiple integer DH groups. The shares for each ClientKeyShareOffer
-MUST by generated independently. Clients MUST NOT offer multiple
-ClientKeyShareOffers for the same parameters. It is explicitly
-permitted to send an empty "client_key_share" extension as this is used
-to elicit the server's parameters if the client has no useful
-information.
+server_offer
+: A single KeyShare value for the negotiated cipher suite.
+  Servers MUST NOT send a KeyShare for a group not offered
+  by the client.
+{:br }
 
-[[TODO: Recommendation about what the client offers.
-Presumably which integer DH groups and which curves.]]
+Servers offer exactly one KeyShare value, which corresponds to the
+key exchange used for the negotiated cipher suite.
+
+Clients offer an arbitrary number of KeyShare values, each representing
+a single set of key exchange parameters. For instance, a client might
+offer shares for several elliptic curves or multiple integer DH groups.
+The key_exchange values for each KeyShare MUST by generated independently.
+Clients MUST NOT offer multiple KeyShare values for the same parameters.
+Clients MAY omit this extension from the ClientHello, and in response to this,
+servers MUST send a HelloRetryRequest requesting use of one of the
+groups the client offered support for in its "supported_groups"
+extension. If no common supported group is available, the server MUST
+produce a fatal "handshake_failure" alert. (see {{hello-retry-request}})
+
+Clients are RECOMMENDED to send a selection of KeyShare offers to
+avoid the round trip time incurred with a HelloRetryRequest. A
+general RECOMMENDED set of KeyShare offers is as follows, in
+descending priority: X25519, secp256r1, & ffdhe3072. Note that
+clients MUST NOT offer KeyShare values for key exchange methods
+not specified in the set of cipher suites offered in its ClientHello.
+(e.g. the recommendation of ffdhe3072 is not applicable if not
+offering any DHE cipher suites) Clients with prior expectations of
+what groups are likely to be supported by the server will generally
+want to include those groups in addition to, or instead of, this
+general set. Groups stated as mandatory to implement for ciphers
+offered by the client are RECOMMENDED to be sent in most, if not
+all, offers. (see {{mti-cipher-suites}})
```

Comment:
Well, the CFRG was requested to provide new curves because we're not satisfied with the ones we have, so putting the newer and better performing one that was requested first isn't exactly a stretch. (especially because it's only a general recommendation) I thought this was generic enough, but I'll cut it out and treat it as a different issue.